### PR TITLE
bugfix: escape http header

### PR DESCRIPTION
**Pull request by zhuizhuhaomeng**
bugfix:
    1. escape \r\n\0 in http header value, escape CTLs or separators in http header name
    2. allow \t in http uri
    
    goto rfc7230 https://tools.ietf.org/html/rfc7230#section-3.2 for detail

I hereby granted the copyright of the changes in this pull request
to the authors of this lua-nginx-module project.

### CODE REVIEW

File: src/ngx_http_lua_util.c [hunk 1]
```diff
@@ -4327,5 +4364,36 @@ ngx_http_lua_escape_log(u_char *dst, u_char *src, size_t size)
     return 0;
 }
 
+ngx_int_t ngx_http_lua_copy_escape_header(ngx_http_request_t *r,
```
Comment (agentzh) [inline src/ngx_http_lua_util.c:4367]: Bad coding style. Forgot to run the `ngx-releng` tool?

File: src/ngx_http_lua_util.c [hunk 2]
```diff
@@ -4327,5 +4364,36 @@ ngx_http_lua_escape_log(u_char *dst, u_char *src, size_t size)
...
+    ngx_str_t *dst, ngx_int_t header_name)
```
Comment (agentzh) [inline src/ngx_http_lua_util.c:4368]: What does this `header_name` argument mean? If it's a boolean value, then we should use the type `unsigned` and parameter name `is_name` instead.

File: src/ngx_http_lua_util.c [hunk 3]
```diff
@@ -4327,5 +4364,36 @@ ngx_http_lua_escape_log(u_char *dst, u_char *src, size_t size)
...
+        dst->data = ngx_palloc(r->pool, len + 2 * escape + 1);
+        if (dst->data == NULL) {
+            return NGX_ERROR;
+        }
+
+        ngx_http_lua_escape_uri(dst->data, data, len, type);
+        dst->len = len + 2 * escape;
+        dst->data[dst->len] = '\0';
+    }
+    return NGX_OK;
```
Comment (agentzh) [inline src/ngx_http_lua_util.c:4396]: Please always use a blank line after `if` statement code blocks. We mentioned this before.

File: src/ngx_http_lua_util.c [hunk 4]
```diff
@@ -4327,5 +4364,36 @@ ngx_http_lua_escape_log(u_char *dst, u_char *src, size_t size)
...
+         * header values must be null-terminated
+         * */
+        dst->data = ngx_palloc(r->pool, len + 2 * escape + 1);
+        if (dst->data == NULL) {
+            return NGX_ERROR;
+        }
+
+        ngx_http_lua_escape_uri(dst->data, data, len, type);
+        dst->len = len + 2 * escape;
+        dst->data[dst->len] = '\0';
```
Comment (agentzh) [inline src/ngx_http_lua_util.c:4394]: Do we really need the `\0` byte here? If yes, why? Better leave a code comment here to explain.

File: t/113-req-header-cookie.t [hunk 5]
```diff
@@ -263,9 +263,10 @@ Cookie: boo=123; foo=bar
...
-unsafe byte "0xa" in header value "boo=123\x0Afoo"
-failed to set header
+--- response_body
+Cookie foo: bar%0Dbar
+Cookie baz: 
+Cookie boo: 123%0Afoo
+Cookie: boo=123%0Afoo; foo=bar%0Dbar
 --- no_error_log
-[crit]
+[error]
```
Comment (agentzh) [inline t/113-req-header-cookie.t:272]: Please always add one and only one newline at the end of the test file (and other source files).

File: src/ngx_http_lua_util.c (hunk 1 above)
Comment (zhuizhuhaomeng) [reply src/ngx_http_lua_util.c]: ngx-releng won't complain when return type and function name at the same line

File: src/ngx_http_lua_util.c (hunk 4 above)
Comment (zhuizhuhaomeng) [reply src/ngx_http_lua_util.c:4395]: there is a comment on line 4384

File: src/ngx_http_lua_headers_in.c [hunk 6]
```diff
@@ -653,16 +653,19 @@ ngx_http_lua_set_input_header(ngx_http_request_t *r, ngx_str_t key,
...
     ngx_uint_t                        i;
 
     dd("set header value: %.*s", (int) value.len, value.data);
 
-    if (ngx_http_lua_check_unsafe_string(r, key.data, key.len,
-                                         "header name") != NGX_OK
-        || ngx_http_lua_check_unsafe_string(r, value.data, value.len,
-                                            "header value") != NGX_OK)
-    {
+    rc = ngx_http_lua_copy_escape_header(r, &key, 1);
```
Comment (agentzh) [inline src/ngx_http_lua_headers_in.c:662]: I think "copy escaped header" sounds more correct than "copy escape header". In English we cannot just put two verbs together...

File: src/ngx_http_lua_util.c [hunk 7]
```diff
@@ -4327,5 +4364,38 @@ ngx_http_lua_escape_log(u_char *dst, u_char *src, size_t size)
     return 0;
 }
```
Comment (agentzh) [inline src/ngx_http_lua_util.c:4366]: Style: 2 blank lines are needed here. Forgot to run `ngx-releng` again?

File: src/ngx_http_lua_util.c [hunk 8]
```diff
@@ -4327,5 +4364,38 @@ ngx_http_lua_escape_log(u_char *dst, u_char *src, size_t size)
...
+
+    data = dst->data;
+    len = dst->len;
+
+    escape = ngx_http_lua_escape_uri(NULL, data, len, type);
+    if (escape > 0) {
+        /*
+         * we allocate space for the trailling '\0' char here because nginx
+         * header values must be null-terminated
+         * */
```
Comment (agentzh) [inline src/ngx_http_lua_util.c:4387]: Style: `* */` should be `*/`.

File: src/ngx_http_lua_util.c (hunk 4 above)
Comment (agentzh) [reply src/ngx_http_lua_util.c:4395]: Gotcha.

File: src/ngx_http_lua_util.h [hunk 9]
```diff
@@ -27,6 +27,9 @@
 #   define NGX_HTTP_SWITCHING_PROTOCOLS 101
 #endif
 
+#define NGX_HTTP_LUA_ESCAPE_HEADER_NAME  7
+
+#define NGX_HTTP_LUA_ESCAPE_HEADER_VALUE 8
```
Comment (agentzh) [inline src/ngx_http_lua_util.h:32]: Please use 2 spaces before `8`, just like the previous `#define` on line 30.

File: src/ngx_http_lua_headers_out.c [hunk 10]
```diff
@@ -488,14 +488,17 @@ ngx_http_lua_set_output_header(ngx_http_request_t *r, ngx_http_lua_ctx_t *ctx,
...
 
     dd("set header value: %.*s", (int) value.len, value.data);
 
-    if (ngx_http_lua_check_unsafe_string(r, key.data, key.len,
-                                         "header name") != NGX_OK
-        || ngx_http_lua_check_unsafe_string(r, value.data, value.len,
-                                            "header value") != NGX_OK)
-    {
+    rc = ngx_http_lua_copy_escape_header(r, &key, 1);
+    if (rc != NGX_OK) {
```
Comment (agentzh) [inline src/ngx_http_lua_headers_out.c:496]: We don't need to introduce the `rc` variable here since it's always used only once. So better be:

```C
if (ngx_http_lua_copy_escape_header(r, &key, 1) != NGX_OK) {
    return NGX_ERROR;
}
```

The same applies to the `if` statement below.

File: src/ngx_http_lua_util.h [hunk 11]
```diff
@@ -169,6 +172,9 @@ void ngx_http_lua_unescape_uri(u_char **dst, u_char **src, size_t size,
 uintptr_t ngx_http_lua_escape_uri(u_char *dst, u_char *src,
     size_t size, ngx_uint_t type);
 
+ngx_int_t ngx_http_lua_copy_escaped_header(ngx_http_request_t *r,
+    ngx_str_t *dst, ngx_uint_t header_name);
```
Comment (agentzh) [inline src/ngx_http_lua_util.h:176]: We usually use `int` or `unsigned` for boolean types (the only exception is in the nginx directive configuration where we use `ngx_flag_t` for boolean config values). So `int` here is better for `is_name` (BTW, it should be `is_name` than `header_name`).

`ngx_uint_t` is like `unsigned long` on x64, BTW, which is 8 bytes long.

File: src/ngx_http_lua_util.c [hunk 12]
```diff
@@ -4328,4 +4365,38 @@ ngx_http_lua_escape_log(u_char *dst, u_char *src, size_t size)
 }
 
 
+ngx_int_t
+ngx_http_lua_copy_escaped_header(ngx_http_request_t *r,
+    ngx_str_t *dst, ngx_uint_t is_name)
```
Comment (agentzh) [inline src/ngx_http_lua_util.c:4370]: Like below.

File: src/ngx_http_lua_util.c [hunk 13]
```diff
@@ -2008,8 +2008,45 @@ ngx_http_lua_escape_uri(u_char *dst, u_char *src, size_t size, ngx_uint_t type)
 
                     /* mail_auth is the same as memcached */
```
Comment (agentzh) [inline src/ngx_http_lua_util.c:2010]: We should summary here in a comment about exactly what characters are to be escaped, like `/* " ", "%", %00-%1F */` for the memcached case above.

File: src/ngx_http_lua_util.c [hunk 14]
```diff
@@ -2008,8 +2008,45 @@ ngx_http_lua_escape_uri(u_char *dst, u_char *src, size_t size, ngx_uint_t type)
...
 
+    static uint32_t header_name[] = {
```
Comment (agentzh) [inline src/ngx_http_lua_util.c:2011]: Style: needs 2 spaces before the identifier `header_name`. See above for examples (like `memcached[]`).

File: src/ngx_http_lua_util.c [hunk 15]
```diff
@@ -2008,8 +2008,45 @@ ngx_http_lua_escape_uri(u_char *dst, u_char *src, size_t size, ngx_uint_t type)
...
+                    /*  ~}| {zyx wvut srqp  onml kjih gfed cba` */
+        0xa8000000, /* 1010 1000 0000 0000  0000 0000 0000 0000 */
+
+        0x00000000, /* 0000 0000 0000 0000  0000 0000 0000 0000 */
+        0x00000000, /* 0000 0000 0000 0000  0000 0000 0000 0000 */
+        0x00000000, /* 0000 0000 0000 0000  0000 0000 0000 0000 */
+        0x00000000, /* 0000 0000 0000 0000  0000 0000 0000 0000 */
+    };
+
+    static uint32_t header_value[] = {
```
Comment (agentzh) [inline src/ngx_http_lua_util.c:2029]: Ditto.

File: src/ngx_http_lua_util.h [hunk 16]
```diff
@@ -500,7 +506,7 @@ ngx_http_lua_check_unsafe_string(ngx_http_request_t *r, u_char *str, size_t len,
                      /* %00-%1F, %7F */
```
Comment (agentzh) [inline src/ngx_http_lua_util.h:506]: You forgot the update this comment to reflect the change below?

I believe a general "unsafe string" makes no sense (as I mentioned in a review comment yesterday). Headers, URIs, URI arguments may have different "unsafe" byte ranges anyway. We'd better be more explicit about our intention.

BTW, do we really need this function now?

File: t/016-resp-header.t [hunk 17]
```diff
@@ -1976,14 +1976,12 @@ Content-Type: application/json
     }
 --- request
 GET /t
---- error_code: 500
 --- response_headers
-header:
+header: value%0Dfoo:bar%0Abar:foo
```
Comment (agentzh) [inline t/016-resp-header.t:1980]: Do we need to update the documentation to reflect the behavioral changes in this PR?

File: src/ngx_http_lua_util.h [hunk 18]
```diff
@@ -500,7 +506,7 @@ ngx_http_lua_check_unsafe_string(ngx_http_request_t *r, u_char *str, size_t len,
...
 
     static uint32_t  unsafe[] = {
-        0xffffffff, /* 1111 1111 1111 1111  1111 1111 1111 1111 */
+        0xfffffdff, /* 1111 1111 1111 1111  1111 1101 1111 1111 */
```
Comment (agentzh) [inline src/ngx_http_lua_util.h:509]: Why this change?

File: src/ngx_http_lua_headers_in.c [hunk 19]
```diff
@@ -653,16 +653,19 @@ ngx_http_lua_set_input_header(ngx_http_request_t *r, ngx_str_t key,
 {
     ngx_http_lua_header_val_t         hv;
     ngx_http_lua_set_header_t        *handlers = ngx_http_lua_set_handlers;
+    ngx_int_t                         rc;
```
Comment (agentzh) [inline src/ngx_http_lua_headers_in.c:657]: We can remove this blank line, BTW.

File: src/ngx_http_lua_util.c [hunk 20]
```diff
@@ -4328,4 +4365,38 @@ ngx_http_lua_escape_log(u_char *dst, u_char *src, size_t size)
...
+        ? NGX_HTTP_LUA_ESCAPE_HEADER_NAME : NGX_HTTP_LUA_ESCAPE_HEADER_VALUE;
+
+    data = dst->data;
+    len = dst->len;
+
+    escape = ngx_http_lua_escape_uri(NULL, data, len, type);
+    if (escape > 0) {
+        /*
+         * we allocate space for the trailling '\0' char here because nginx
+         * header values must be null-terminated
```
Comment (agentzh) [inline src/ngx_http_lua_util.c:4387]: Will you add a reference to this statement? Like "see src/blah/blah.c" for details.

File: README.markdown [hunk 21]
```diff
@@ -4883,6 +4883,11 @@ ngx.req.set_header
 
 Set the current request's request header named `header_name` to value `header_value`, overriding any existing ones.
 
+The input Lua string header_name and header_value should already be well-formed with the URI encoding.
```
Comment (agentzh) [inline README.markdown:4886]: Missing backticks around `header_name` and `header_value`. See the previous paragraph for examples.

File: README.markdown [hunk 22]
```diff
@@ -4883,6 +4883,11 @@ ngx.req.set_header
...
+For security considerations, this method will automatically escape " ", """, "(", ")", ",", "/", ":", ";", "?",
+"<", "=", ">", "?", "@", "[", "]", "\", "{", "}", %00-%1F, %7F-%FF in header_name and automatically escape
+"%00-%08, %0A-%0F, %7F in header value.
```
Comment (agentzh) [inline README.markdown:4889]: Better use `0xFF` syntax instead of the URI-encoded form.
